<a href="https://colab.research.google.com/github/bala-srm/hello-world/blob/master/RandomForest_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from google.colab import files
from sklearn.model_selection import train_test_split as tts
from sklearn.metrics import accuracy_score

In [3]:
files.upload()

Saving data_cleaned.csv to data_cleaned.csv


{'data_cleaned.csv': b'Survived,Age,Fare,Pclass_1,Pclass_2,Pclass_3,Sex_female,Sex_male,SibSp_0,SibSp_1,SibSp_2,SibSp_3,SibSp_4,SibSp_5,SibSp_8,Parch_0,Parch_1,Parch_2,Parch_3,Parch_4,Parch_5,Parch_6,Embarked_C,Embarked_Q,Embarked_S\n0,22.0,7.25,0,0,1,0,1,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n1,38.0,71.2833,1,0,0,1,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0\n1,26.0,7.925,0,0,1,1,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n1,35.0,53.1,1,0,0,1,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n0,35.0,8.05,0,0,1,0,1,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n0,29.69911764705882,8.4583,0,0,1,0,1,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0\n0,54.0,51.8625,1,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n0,2.0,21.075,0,0,1,0,1,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,1\n1,27.0,11.1333,0,0,1,1,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1\n1,14.0,30.0708,0,1,0,1,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0\n1,4.0,16.7,0,0,1,1,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1\n1,58.0,26.55,1,0,0,1,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1\n0,20.0,8.05,0,0,1,0,1,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0

In [4]:
df = pd.read_csv('data_cleaned.csv')

In [5]:
df.shape

(891, 25)

In [6]:
df.head()

Survived   Age     Fare  ...  Embarked_C  Embarked_Q  Embarked_S
0         0  22.0   7.2500  ...           0           0           1
1         1  38.0  71.2833  ...           1           0           0
2         1  26.0   7.9250  ...           0           0           1
3         1  35.0  53.1000  ...           0           0           1
4         0  35.0   8.0500  ...           0           0           1

[5 rows x 25 columns]

### Split features into Target and Independent Features

In [7]:
x = df.drop('Survived', axis = 1)
y = df.Survived

### Split the Dataset into Train and Test Dataset

In [9]:
train_x, test_x, train_y, test_y = tts(x,y, random_state = 123, stratify = y)

In [11]:
train_x.shape, test_x.shape

((668, 24), (223, 24))

### Random Forest Classifier Implementation

In [12]:
RF = RandomForestClassifier(random_state=123)

In [13]:
RF.fit(train_x,train_y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [14]:
rf_train_pred = RF.predict(train_x)

In [15]:
RF.score(train_x,train_y)

0.9865269461077845

In [19]:
train_acc = accuracy_score(rf_train_pred, train_y)
train_acc

0.9865269461077845

In [16]:
rf_test_pred = RF.predict(test_x)

In [20]:
RF.score(test_x,test_y)

0.7757847533632287

In [21]:
test_acc = accuracy_score(rf_test_pred, test_y)
test_acc

0.7757847533632287

### Tweak the Hyperparameters of the Random Forest Model

In [22]:
RF2 = RandomForestClassifier(n_estimators=200,
                             max_depth =5,
                             min_samples_split = 25,
                             min_samples_leaf = 25)

In [23]:
RF2.fit(train_x,train_y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=5, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=25, min_samples_split=25,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [24]:
RF2.score(train_x,train_y)

0.8023952095808383

In [25]:
RF2.score(test_x,test_y)

0.7668161434977578

In [32]:
n_trees = [50,100,150,200]
max_feat = [3,5,7]
max_dep = [3,4,5,6]
min_samples_split = [10,20,30,40]

In [33]:
acc_score = []
trees = []
dep = []
min_split =[]
max_features = []

def rf_iterate(n_trees, max_feat, max_dep, min_samples_split):
  
  for i in n_trees:
    for j in max_feat:
      for k in max_dep:
        for l in min_samples_split:
          RF3 = RandomForestClassifier(n_estimators=i,
                               max_depth = k,
                               max_features = j,
                               min_samples_split = l,
                               random_state = 123)
          RF3.fit(train_x,train_y)
          acc_score.append(RF3.score(test_x,test_y))
          trees.append(i)
          dep.append(k)
          min_split.append(l)
          max_features.append(j)


In [34]:
rf_iterate(n_trees, max_feat,max_dep,min_samples_split)

In [35]:
perf_df = pd.DataFrame({'Trees':trees,
                        'Depth' : dep,
                        'Min_Sam_Spl' : min_split,
                        'Max_Features_spl' :max_features,
                        'Accuracy' : acc_score
    
})

In [36]:
perf_df[perf_df.Accuracy==perf_df.Accuracy.max()]

Trees  Depth  Min_Sam_Spl  Max_Features_spl  Accuracy
138    150      5           30                 7  0.825112
186    200      5           30                 7  0.825112

So we can select a model with either of the above 2 sets of parameters and improve the accuracy score of the model to 0.825 as opposed to 0.78 or 0.80 with earlier models.